engine.js

```diff
@@ -737,37 +737,83 @@ var accumulateResult = function(ifResult, thenResult) {
+            .then(function(response) {
+                var result = resultFromResponse(response);
+                if (result.result) {
+                    if ((element.metroArea === 'NA' && result.msa_code!=='') || element.metroArea!==result.msa_code) {
+                        var error = {'properties': {}};
+                        error.properties['Recommended MSA/MD'] = result.msa_code;
+                        error.properties['LAR number'] = element.loanNumber;
+                        error.properties['Reported State Code'] = element.fipsState;
+                        error.properties['Reported County Code'] = element.fipsCounty;
+                        error.properties['Reported Census Tract'] = element.censusTract;
+                        invalidMSAs.push (error);
+                    }
                 }
+                return Promise.resolve();
+            });
+        }, { concurrency: CONCURRENT_RULES })
+        .then(function() {
+            if (!invalidMSAs.length) {
                 return true;
+            } else {
+                return invalidMSAs;
             }
-            return false;
-        });
-    };
-
-    HMDAEngine.isNotIndependentMortgageCoOrMBS = function(respondentID, agencyCode) {
-        return this.apiGET('isNotIndependentMortgageCoOrMBS', [agencyCode, respondentID])
-        .then(function(response) {
-            return resultFromResponse(response).result;
         });
     };
 
-    HMDAEngine.isMetroAreaOnRespondentPanel = function(metroArea, respondentID, agencyCode) {
-        return this.apiGET('isMetroAreaOnRespondentPanel', [agencyCode, respondentID, metroArea])
-        .then(function(response) {
-            return resultFromResponse(response).result;
+    HMDAEngine.isMetroAreaOnRespondentPanel = function(hmdaFile) {
+        var currentEngine = this,
+            invalidMSAs = {};
+        return Promise.map(hmdaFile.loanApplicationRegisters, function(element) {
+            return currentEngine.apiGET('isNotIndependentMortgageCoOrMBS', [element.agencyCode, element.respondentID])
+            .then(function(response) {
+                var result = resultFromResponse(response);
+                if (result.result) {
+                    var validActionTaken = ['1', '2', '3', '4', '5', '7', '8'];
+                    if (_.contains(validActionTaken, element.actionTaken)) {
+                        return currentEngine.apiGET('isMetroAreaOnRespondentPanel', [element.agencyCode, element.respondentID,
+                                            element.metroArea])
+                        .then(function(response) {
+                            var result = resultFromResponse(response);
+                            if (!result.result) {
+                                if (invalidMSAs[element.metroArea] !== undefined) {
+                                    invalidMSAs[element.metroArea]['LAR Count'] ++;
```

Comment:
It is strange to see a space between a variable and a unary operator. This is supposed to be `++` and not `+ x`, right?